MQE-1886: Ability to use grab data between ActionGroups

soumyau opened the pull request:
<!--- Provide a general summary of the Pull Request in the Title above -->

### Description
<!--- Provide a description of the changes proposed in the pull request -->

### Fixed Issues (if relevant)
<!--- Provide a list of fixed issues in the format magento/magento2-functional-testing-framework#<issue_number>, if relevant  -->
1. magento/magento2-functional-testing-framework#<issue_number>: Issue title
2. ...

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/verification tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)
 - [ ] Changes to Framework doesn't have backward incompatible changes for tests or have related Pull Request with fixes to tests

jilu1 commented on src/Magento/FunctionalTestingFramework/Test/etc/mergedActionGroupSchema.xsd:
```diff
@@ -15,29 +15,21 @@
+                    <xs:element ref="annotations"/>
+                </xs:choice>
+            </xs:sequence>
+        </xs:sequence>
```
In case `return` is forced to be the last action in an action group, this probably can be simplified.

soumyau replied:
when `return` is forced to be the last step, it throws a schema validation error in case of merging actionGroups, based on order of reading files.

```
 <actionGroup name="MergeActionGroup">
           <arguments>
               <argument name="value" type="string"/>
        </arguments>
        <fillField selector="#foo" userInput="{{value}}" stepKey="fill1"/>
        <grabTextFrom selector="#bar" stepKey="grab1"/>
     
</actionGroup>
 <actionGroup name="MergeActionGroup">
      <fillField selector="#foo1" userInput="xyz" stepKey="fill2" after="fill1"/>
      <return value="{$grab1}" stepKey="return2"/>
</actionGroup>

```
Schema validation is run on a sort-of merged file, however, the stepKey order is not resolved at this stage -
```
<actionGroup name="MergeActionGroup" filename="/Users/sunnikri/workspace/magento2ce/app/code/Magento/AdminAnalytics/Test/Mftf/ActionGroup/MergeActionGroup.xml,/Users/sunnikri/workspace/magento2ce/app/code/Magento/Backend/Test/Mftf/ActionGroup/MergeActionGroup.xml">
      <fillField selector="#foo1" userInput="xyz" stepKey="fill2" after="fill1"/>
      <return value="{$grab1}" stepKey="return2"/>
           <arguments>
               <argument name="value" type="string"/>
        </arguments>
        <fillField selector="#foo" userInput="{{value}}" stepKey="fill1"/>
        <grabTextFrom selector="#bar" stepKey="grab1"/>
</actionGroup>
```
`return1` appears before `fill1`, and schema validation fails.

jilu1 replied:
It's fine if this cannot be simplfied.

jilu1 commented on docs/test/action-groups.md:
````diff
@@ -180,6 +180,34 @@ MFTF resolves `{{myCustomEntity.field1}}` the same as it would in a `selector` o
+
+Action groups can return a value using `<return>` action.
+
+```xml
````
The example below and the example in actions.md seem to be duplicate. I would keep one and make a link to the other. But maybe check with @dobooth

soumyau replied:
fixed.

soumyau commented:
> Simple case works as expected. I would suggest to discuss with team to clearly define cases like:
> 
> * should we allow `return` be used in any where in an action group?
please check my comment on `actionGroupSchema.xsd`, looks like this is not feasible in the current setup.
> * what's the expected behaviors when action groups with returns get merged? Currently it results an error
Added documentation on use of return in merging action groups, please check.
> * what's the expected behaviors when action groups with returns get extends? Currently it's results is not ideal and sometimes an error
for extending action groups if return is used in both parent and child, if parent action group is referenced, it returns parent's value. If child action group is ref it returns child value. Seems logical.
> * Make sure to check the `return` in before/after case, it seems not working for me
Added documentation regarding scope.
> * Have we talked about returning more than one values in an array?
<return> can return multiple values if grabMultiple stepKey is referenced. The use case seems rare, if there are users requesting it, we can incorporate later?

soumyau commented:
> And please add verification tests for merge, extends, return values in before/after, etc
added verification steps

jilu1 commented on dev/tests/verification/TestModule/Test/ActionGroupFunctionalTest/ActionGroupReturningValueTest.xml:
```diff
@@ -0,0 +1,30 @@
+            <actionGroup ref="FunctionalActionGroup" stepKey="afterGroup"/>
+        </after>
+    </test>
+</tests>
```
Not need to keep both DevDocs Functional test and Verification test. One should be enough for return testt

soumyau replied:
Removed functional test.

jilu1 commented:
> > Simple case works as expected. I would suggest to discuss with team to clearly define cases like:
> > 
> > * should we allow `return` be used in any where in an action group?
> >   please check my comment on `actionGroupSchema.xsd`, looks like this is not feasible in the current setup.
> > * what's the expected behaviors when action groups with returns get merged? Currently it results an error
> >   Added documentation on use of return in merging action groups, please check.
> > * what's the expected behaviors when action groups with returns get extends? Currently it's results is not ideal and sometimes an error
> >   for extending action groups if return is used in both parent and child, if parent action group is referenced, it returns parent's value. If child action group is ref it returns child value. Seems logical.
> > * Make sure to check the `return` in before/after case, it seems not working for me
> >   Added documentation regarding scope.
> > * Have we talked about returning more than one values in an array?
> >    can return multiple values if grabMultiple stepKey is referenced. The use case seems rare, if there are users requesting it, we can incorporate later?

Do you mean currently we can return multiple values?
If not, let's document it that currently only return one value

Above issues, some of the issues are caused by existing bug (merge) or existing limitations (in before/after). Can we document the expect behavior for them? e.g.
One `return` is allowed among all merging action groups

soumyau commented on docs/test/action-groups.md:
````diff
@@ -180,6 +180,34 @@ MFTF resolves `{{myCustomEntity.field1}}` the same as it would in a `selector` o
+    <argument name="orderId" value="{$getOrderId}"/>
+</actionGroup>
+```
+### Convention to return a value 
````
@jilu1 did you check this documentation. Does this answer the question regarding multiple returns and merging action groups?

jilu1 replied:
Yes. It looks good.
